src/snowflake/connector/storage_client_async.py

```diff
@@ -0,0 +1,302 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import asyncio
+import ssl
+
+import aiohttp
+
+from .network_async import EventLoopThreadRunner
+from .storage_client import *
+
+# YICHUAN: Not much to explain in this file, we add async versions of existing private methods and override public
+# methods to call async private methods instead
+# We explicitly add new private methods with suffix "_async" for clarity instead of overriding
+
+
+class SnowflakeStorageClientAsync(SnowflakeStorageClient):
+    TRANSIENT_ERRORS_ASYNC = (
+        ssl.SSLSyscallError,
+        asyncio.TimeoutError,
+        aiohttp.ClientConnectionError,
+    )
+
+    @property
+    def _loop_runner(self) -> EventLoopThreadRunner:
+        if self.meta.sfagent and self.meta.sfagent._cursor.connection:
+            return self.meta.sfagent._cursor.connection._rest._loop_runner
+        else:
+            return self._loop_runner
+
+    @abstractmethod
+    async def get_file_header_async(self, filename: str) -> FileHeader | None:
+        pass
+
+    def preprocess(self) -> None:
```

Comment:
for future reference: we can convert this method to async once we push async to upper module, specifically we update `self._loop_runner.run_coro` to `await`